cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/partitioned/DynamicPartitioningOutputFormat.java

```diff
@@ -0,0 +1,189 @@
+    final Partitioning partitioning = outputDataset.getPartitioning();
+
+    dynamicPartitioner.initialize(mrTaskContext);
+
+    return new RecordWriter<K, V>() {
+
+      // a cache storing the record writers for different output files.
+      TreeMap<String, RecordWriter<K, V>> recordWriters = new TreeMap<>();
+
+      public void write(K key, V value) throws IOException, InterruptedException {
+        PartitionKey partitionKey = dynamicPartitioner.getPartitionKey(key, value);
+        String relativePath = PartitionedFileSetDataset.getOutputPath(partitionKey, partitioning);
+        String finalPath = relativePath + "/" + outputName;
+
+        RecordWriter<K, V> rw = this.recordWriters.get(finalPath);
+        if (rw == null) {
+          // if we don't have the record writer yet for the final path, create one and add it to the cache
+          rw = getBaseRecordWriter(getTaskAttemptContext(job, finalPath));
+          this.recordWriters.put(finalPath, rw);
+        }
+        rw.write(key, value);
+      }
+
+      @Override
+      public void close(TaskAttemptContext context) throws IOException, InterruptedException {
+        try {
+          for (RecordWriter<K, V> recordWriter : this.recordWriters.values()) {
+            recordWriter.close(context);
+          }
+          this.recordWriters.clear();
+
+          mrTaskContext.flushOperations();
+        } catch (Exception e) {
+          throw new IOException(e);
+        } finally {
+          mrTaskContext.close();
+        }
+      }
+    };
+  }
+
+  private static TaskAttemptContext getTaskAttemptContext(TaskAttemptContext context,
+                                                          String newOutputName) throws IOException {
+    Job job = new Job(context.getConfiguration());
+    FileOutputFormat.setOutputName(job, newOutputName);
+
+    Path jobOutputPath = createJobSpecificPath(FileOutputFormat.getOutputPath(job), context);
+    FileOutputFormat.setOutputPath(job, jobOutputPath);
+
+    return new TaskAttemptContextImpl(job.getConfiguration(), context.getTaskAttemptID());
+  }
+
+  /**
+   * @return A RecordWriter object over the given file
+   * @throws IOException
+   */
+  protected RecordWriter<K, V> getBaseRecordWriter(TaskAttemptContext job) throws IOException, InterruptedException {
+    if (fileOutputFormat == null) {
+      // TODO: use the PFS's OutputFormat (encode in conf)
+      fileOutputFormat = new TextOutputFormat<>();
+    }
+    return fileOutputFormat.getRecordWriter(job);
+  }
+
+  // suffixes a Path with a job-specific string
+  private static Path createJobSpecificPath(Path path, JobContext jobContext) {
+    String outputPathSuffix = "_temporary_" + jobContext.getJobID().getId();
+    return new Path(path, outputPathSuffix);
+  }
+
+  @Override
+  public synchronized FileOutputCommitter getOutputCommitter(TaskAttemptContext context) throws IOException {
+    if (committer == null) {
```

Comment:
Does this have to be through an instance variable? That is, will this ever be called multiple times? And if so, does it matter whether it returns the same instance?